modules/collect/src/main/java/com/opengamma/strata/collect/io/XmlElement.java

```diff
@@ -0,0 +1,391 @@
+   */
+  public XmlElement getChildSingle(String childName) {
+    return getChildOptional(childName)
+        .orElseThrow(() -> new IllegalArgumentException(Messages.format(
+            "Unknown element '{}' in element '{}'", childName, name)));
+  }
+
+  /**
+   * Gets the first matching child element with the specified name.
+   * <p>
+   * This returns the first child element with the specified name.
+   * An exception is thrown if the child does not exist.
+   * 
+   * @param childName  the name to match
+   * @return the child matching the name
+   * @throws IllegalArgumentException if there is no match
+   */
+  public XmlElement getChildFirst(String childName) {
+    return streamChildren(childName)
+        .findFirst()
+        .orElseThrow(() -> new IllegalArgumentException(Messages.format(
+            "Unknown element '{}' in element '{}'", childName, name)));
+  }
+
+  /**
+   * Gets the child elements matching the specified name.
+   * <p>
+   * This returns all the child elements with the specified name.
+   * 
+   * @param childName  the name to match
+   * @return the children matching the name
+   */
+  public ImmutableList<XmlElement> getChildren(String childName) {
+    return streamChildren(childName).collect(toImmutableList());
+  }
+
+  /**
+   * Gets the child elements matching the specified name.
+   * <p>
+   * This returns all the child elements with the specified name.
+   * 
+   * @param childName  the name to match
+   * @return the children matching the name
+   */
+  public Stream<XmlElement> streamChildren(String childName) {
+    return children.stream().filter(child -> child.getName().equals(childName));
+  }
+
+  //-------------------------------------------------------------------------
+  @Override
+  public MetaBean metaBean() {
+    return META_BEAN;
+  }
+
+  @Override
+  public <R> Property<R> property(String propertyName) {
+    return metaBean().<R>metaProperty(propertyName).createProperty(this);
+  }
+
+  @Override
+  public Set<String> propertyNames() {
+    return metaBean().metaPropertyMap().keySet();
+  }
+
+  //-------------------------------------------------------------------------
+  /**
+   * Checks if this element equals another.
+   * <p>
+   * This compares the entire state of the elements.
+   * 
+   * @param obj  the other element, null returns false
+   * @return true if equal
+   */
+  @Override
+  public boolean equals(Object obj) {
+    if (obj == this) {
+      return true;
+    }
+    if (obj instanceof XmlElement) {
+      XmlElement other = (XmlElement) obj;
+      return name.equals(other.name) &&
+          Objects.equals(content, other.content) &&
+          attributes.equals(other.attributes) &&
+          children.equals(other.children);
+    }
+    return false;
+  }
+
+  /**
+   * Returns a suitable hash code.
+   * <p>
+   * Note that the has code excludes the children to ensure performance is reasonable in large trees.
```

Comment:
I'm still not comfortable with breaking such a fundamental rule without a concrete reason. It feels like a premature optimisation.

If we don't think these will be used as map keys then what harm does it do to implement `hashCode` properly? Or not implement it at all? Either would be preferable to a surprising and non-standard implementation that isn't justified by any data.